packages/payment-experience-api/src/api/paytrailOnlineRefundPaymentSuccessController.ts

```diff
@@ -0,0 +1,170 @@
+  protected readonly requestSchema = null
+
+  protected async implementation(
+    request: Request,
+    result: Response
+  ): Promise<any> {
+    const { query } = request
+    // Validates that base redirect url is set
+    PaytrailOnlineRefundPaymentSuccessController.checkAndCreateRedirectUrl()
+    const refundId = parseUuidFromPaytrailRedirectCheckoutStamp({ query })
+
+    if (!refundId) {
+      logger.error(
+        'No refundId specified redirect to paytrail general failure url'
+      )
+      return result.redirect(
+        302,
+        PaytrailOnlineRefundPaymentSuccessController.getFailureRedirectUrl()
+      )
+    }
+
+    const refund = await getRefundAdmin({ refundId })
+    const orderId = refund.refund.orderId
+    const order = await getOrderAdmin({ orderId })
+    const merchantId = parseMerchantIdFromFirstOrderItem(order)
+
+    if (!merchantId) {
+      logger.error('Paytrail: No merchantId found from order')
+      throw new ExperienceError({
+        code: 'merchant-id-not-found',
+        message: 'No merchantId found from order.',
+        responseStatus: StatusCode.NotFound,
+        logLevel: 'info',
+      })
+    }
+
+    try {
+      const paytrailStatus = await checkPaytrailRefundCallbackUrl({
+        params: query,
+        merchantId: merchantId,
+      })
+      logger.debug(
+        `PaytrailStatus for refund ${refundId}: ${JSON.stringify(
+          paytrailStatus
+        )}`
+      )
+      if (!paytrailStatus.valid) {
+        logger.debug(
+          `PaytrailStatus is not valid for refund ${refundId}, redirect to failure url`
+        )
+        return result.redirect(
+          302,
+          PaytrailOnlineRefundPaymentSuccessController.getFailureRedirectUrl()
+        )
+      }
+
+      const redirectUrl = await createUserRedirectUrl({
+        order,
+        paymentReturnStatus: paytrailStatus,
+        redirectPaymentUrlBase: PaytrailOnlineRefundPaymentSuccessController.getRedirectUrl(),
+      })
+      // TODO: Make changes that sends email from refund
+      // Function contains internal checks when to send receipt.
+      // await sendReceiptToCustomer(paytrailStatus, orderId, order)
+
+      // Only cancel authorized card renewals.
+      if (isAuthorized(paytrailStatus) && isCardRenewal(paytrailStatus)) {
+        const cancelled = await this.cancelAuthorizationPayments(order)
+        // Error count more than 0
+        let cancelledCount = cancelled[1]
+        if (cancelledCount != 0) {
+          return result.redirect(
+            302,
+            PaytrailOnlineRefundPaymentSuccessController.getCardRenewalFailureRedirectUrl(
+              order.orderId
+            ).toString()
+          )
+        }
+      }
+
+      return result.redirect(302, redirectUrl.toString())
+    } catch (error) {
+      logger.error(error)
+      logger.debug(
+        `Error occurred, redirect user to failure url for order ${orderId}`
```

Comment:
Change error log to contain unique text and contain refund id, paytrail etc

Reply:
Sure!